rfcs/0001-masonry-backend.md

```diff
@@ -0,0 +1,166 @@
+
+As a result, Xilem's native backend is in a poor state:
+
+- There is [code commented out](https://github.com/linebender/xilem/blob/ea45b9f8c14e3708f0fcbe0a0e1c760f59146323/src/widget/widget.rs#L113-L120).
+- There are [*entire modules* commented out](https://github.com/linebender/xilem/blob/ea45b9f8c14e3708f0fcbe0a0e1c760f59146323/src/widget/mod.rs#L19-L20).
+- There is [documentation referring to items from Druid that no longer exist](https://github.com/linebender/xilem/blob/ea45b9f8c14e3708f0fcbe0a0e1c760f59146323/src/widget/widget.rs#L51-L71).
+- There are [TODOs without an associated issue](https://github.com/linebender/xilem/blob/ea45b9f8c14e3708f0fcbe0a0e1c760f59146323/src/widget/core.rs#L569C5-L569C66).
+
+While Masonry's codebase hasn't been updated in a while, I made sure to address the above problems before I stepped away from it. I would also argue it has a tighter vision and more helpful documentation, thought that is more subjective.
+
+Masonry also comes with some built-in perks, like powerful unit tests and a structured widget graph.
+
+For these reasons, I think a new backend should be added to Xilem using either the Masonry crate or one forked from it. (We will refer to that new backend as "Masonry" in the rest of this RFC, although it might have a different name in the end.)
+
+
+## User-facing explanation
+
+### Xilem architecture
+
+Xilem is roughly separated into two layers:
+
+- **The reactive layer** handles the app logic and centralized state. Your application's code will call functions from this layer.
+- **The platform layer** handles the document state, depending on the platform.
+
+The reactive layer defines the `View` trait as well as other items for sequences, type-erasure and memoization. A `View` represent a transient node or tree of graphical items, which is return by a declarative function we usually call a "component". An example of component could be:
+
+```rust
+fn app_logic(data: &mut u32) -> impl View<u32, (), Element = impl Widget> {
+    Column::new((
+        Button::new(format!("count: {}", data), |data| *data += 1),
+        Button::new("reset", |data| *data = 0),
+    ))
+}
+```
+
+In that example, `app_logic()` returns a `Column` view which contains two `Button` views. These store the minimum information to represent the concept of two buttons stored in a column.
+
+The `View` trait has an `Element` associated type, which depends on the platform layer, and represents a persisten UI entity. Every time app data changes, a new View is produced. That View is diffed against the app's previous View, and all changes are applied to matching elements: if the new view has a child view that wasn't in the previous version, a new element is added; if the new view no longer has a child view that was in the previous version, the matching element is removed, and so on.
```

Comment:
```suggestion
The `View` trait has an associated type `Element`, which depends on the platform layer, and represents a persistent UI entity. Every time app data changes, a new View is produced. That View is diffed against the app's previous View, and all changes are applied to matching elements: if the new view has a child view that wasn't in the previous version, a new element is added; if the new view no longer has a child view that was in the previous version, the matching element is removed, and so on.
```

Not sure whether it "depends on the platform layer" but is just the actual type of the UI Widget (The platform is rather defined by the runtime).
As you say yourself below, the `View` trait is a different type on each platform, requiring different `Widget` traits for the `Element` type
 
Might be nitpicking though...

Reply:
Yeah, I rushed that explanation a bit. Didn't want to have to document the entire xilem_core paradigm just to justify switching a backend.